### PR TITLE
Netpol E2E test should allow traffic to an application

### PR DESCRIPTION
**Pull request by knabben**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Increasing coverage cases for Netpol tests, adding the case to ALLOW all Ingress traffic to an application takes precedence over a DENY all ingress on a namespace. 

**Which issue(s) this PR fixes**:

Refs https://github.com/kubernetes/kubernetes/issues/97640

**Special notes for your reviewer**:
Covers [2a case](https://github.com/ahmetb/kubernetes-network-policy-recipes/blob/master/02a-allow-all-traffic-to-an-application.md) - Allow All traffic to an application.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by knabben** [comment]
```Dec 30 16:59:40.754: INFO: expected:

-               netpol-9038-x/a netpol-9038-x/b netpol-9038-x/c netpol-9038-y/a netpol-9038-y/b netpol-9038-y/c netpol-9038-z/a netpol-9038-z/b netpol-9038-z/c
netpol-9038-x/a .               X               X               .               .               .               .               .               .
netpol-9038-x/b .               X               X               .               .               .               .               .               .
netpol-9038-x/c .               X               X               .               .               .               .               .               .
netpol-9038-y/a .               X               X               .               .               .               .               .               .
netpol-9038-y/b .               X               X               .               .               .               .               .               .
netpol-9038-y/c .               X               X               .               .               .               .               .               .
netpol-9038-z/a .               X               X               .               .               .               .               .               .
netpol-9038-z/b .               X               X               .               .               .               .               .               .
netpol-9038-z/c .               X               X               .               .               .               .               .               .



Dec 30 16:59:40.754: INFO: observed:

-               netpol-9038-x/a netpol-9038-x/b netpol-9038-x/c netpol-9038-y/a netpol-9038-y/b netpol-9038-y/c netpol-9038-z/a netpol-9038-z/b netpol-9038-z/c
netpol-9038-x/a .               X               X               .               .               .               .               .               .
netpol-9038-x/b .               X               X               .               .               .               .               .               .
netpol-9038-x/c .               X               X               .               .               .               .               .               .
netpol-9038-y/a .               X               X               .               .               .               .               .               .
netpol-9038-y/b .               X               X               .               .               .               .               .               .
netpol-9038-y/c .               X               X               .               .               .               .               .               .
netpol-9038-z/a .               X               X               .               .               .               .               .               .
netpol-9038-z/b .               X               X               .               .               .               .               .               .
netpol-9038-z/c .               X               X               .               .               .               .               .               .



Dec 30 16:59:40.754: INFO: comparison:

-               netpol-9038-x/a netpol-9038-x/b netpol-9038-x/c netpol-9038-y/a netpol-9038-y/b netpol-9038-y/c netpol-9038-z/a netpol-9038-z/b netpol-9038-z/c
netpol-9038-x/a .               .               .               .               .               .               .               .               .
netpol-9038-x/b .               .               .               .               .               .               .               .               .
netpol-9038-x/c .               .               .               .               .               .               .               .               .
netpol-9038-y/a .               .               .               .               .               .               .               .               .
netpol-9038-y/b .               .               .               .               .               .               .               .               .
netpol-9038-y/c .               .               .               .               .               .               .               .               .
netpol-9038-z/a .               .               .               .               .               .               .               .               .
netpol-9038-z/b .               .               .               .               .               .               .               .               .
netpol-9038-z/c .               .               .               .               .               .               .               .               .
```

Matrix generated with Cilium 1.9.1

File: test/e2e/network/netpol/network_policy.go [hunk 1]
```diff
@@ -217,6 +186,25 @@ var _ = SIGDescribeCopy("Netpol [LinuxOnly]", func() {
 			ValidateOrFail(k8s, model, &TestCase{FromPort: 81, ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: reachability})
 		})
 
+		ginkgo.It("should enforce policy to allow all traffic to an application, having a deny policy [Feature:NetworkPolicy] ", func() {
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:189]: As I understand it, the point of this test case is to verify that given a deny-all and an explicit allow for the same target, the allow wins -- do I have that right?

Can that be clarified in the test description?  Trying to think of a more elegant way to state that :)

File: test/e2e/network/netpol/network_policy.go [hunk 2]
```diff
@@ -217,6 +186,25 @@ var _ = SIGDescribeCopy("Netpol [LinuxOnly]", func() {
...
+			nsX, _, _, model, k8s := getK8SModel(f)
+
+			// Deny all Ingress traffic policy to pods on nsX
+			policy := GetDenyIngress("deny-all")
+			CreatePolicy(k8s, policy, nsX)
+
+			// Allow Ingress traffic only for x/a from any pod
+			policy = GetAllowIngressForTarget("web-allow-all", map[string]string{"pod": "a"})
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:197]: maybe use a new variable `allowPolicy` to make this easier to read

File: test/e2e/network/netpol/network_policy.go [hunk 3]
```diff
@@ -159,39 +158,9 @@ var _ = SIGDescribeCopy("Netpol [LinuxOnly]", func() {
 		})
 
 		ginkgo.It("should support a 'default-deny-all' policy [Feature:NetworkPolicy]", func() {
-			np := &networkingv1.NetworkPolicy{}
```
Comment (mattfenwick) [inline test/e2e/network/netpol/network_policy.go:162]: these changes seem to be unrelated to the new test case , should they be in a separate PR ?

File: test/e2e/network/netpol/network_policy.go (hunk 2 above)
Comment (mattfenwick) [reply test/e2e/network/netpol/network_policy.go:197]: `allow-all-to-a` would be a more meaningful name for the netpol

**Review by mattfenwick — COMMENTED** [review]
Looks good @knabben !

File: test/e2e/network/netpol/network_policy.go (hunk 1 above)
Comment (knabben) [reply test/e2e/network/netpol/network_policy.go:189]: Maybe moving the having a deny all traffic policy to a Ginkgo.By() clause in the test.
```
Ginkgo.By("having a deny Ingress policy", func() {
   policy = DenyPolicy() 
   CreatePolicy()
    ...
}) 
```

File: test/e2e/network/netpol/network_policy.go (hunk 3 above)
Comment (knabben) [reply test/e2e/network/netpol/network_policy.go:162]: Let me split them

Comment (knabben) [reply test/e2e/network/netpol/network_policy.go]: https://github.com/kubernetes/kubernetes/pull/97638
